### DISCUSSION TITLE
How to run csv-importer behind TLS

### DISCUSSION THREAD

**Discussion by EliRibble** [firefly-iii/firefly-iii#4698] [Firefly III] [answered]
I am running Firefly III version 5.5.8.
I am running csv-importer version

**Description**
I'm attempting to use the web UI for csv-importer behind nginx as a reverse proxy. It doesn't allow regular http requests and redirects to https. The csv-importer UI makes various requests to assets (css, JS) that fail and critically attempts a request to `/token/validate` over http using XHR which fails. I couldn't find any information in the docs for csv-importer to tell it to create URLs with `https://` instead of `http://`. How should we do that?

**Extra info**
 

It looks like csv-importer uses Laravel. I was able to force `https` by modifying `/var/lib/http/app/Providers/RouteServiceProvider.php` and added the following:

```
  public function boot() {
    resolve(\Illuminate\Routing\UrlGenerator::class)->forceScheme('https');
  }
```

This seems pretty heavy-handed. I tried adding all of the following environment variables - none did what I want:

```
BASE_URL=https://importer.my.tld
CSV_IMPORTER_URL=https://importer.my.tld
ASSET_URL=https://importer.my.tld
APP_URL=https://importer.my.tld
FORCE_HTTPS=true
EXPECT_SECURE_URL=1
```

`ASSET_URL` *did* fix asset loading - just not the `/token/validate` request.

**Bonus points**

- I have read the FAQ at https://bit.ly/FF3-FAQ
- I added a screenshot
- I added log files 
- I was able to replicate the issue on the demo site.

**Comment by JC5** [comment]
What's your reverse proxy config?

If you don't configure it correctly it won't work. If you do it right you don't need to change anything in the CSV importer.

**Comment by EliRibble** [comment]
I'm using nginx which responds with a 301 redirect to `https://$host$request_url` any request on port 80.

I'm not aware of any way to get nginx to either modify html/css/JS it serves up. You have to tell the app being proxied to make different requests. At least, as far as I know.

**Reply by JC5** [reply]
```
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header X-Forwarded-Server $host;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header Host $host;
		client_max_body_size 64M;
		proxy_read_timeout 300s;
```

That should do it.

**Reply by EliRibble** [reply]
I already had most of those headers set, with the noted difference that `X-Forwarded-Host` was set to `$http_host` and I didn't have `X-Forwarded-Server` set at all.

That said, adding those headers doesn't fix the issue. I had hoped that the application would use the `X-Forwarded-Proto` header when creating URLs. It does not. Running `phpinfo()` within the container allows me to confirm that `X-Forwarded-Proto` (and the other headers) are properly set and detected by PHP.

**Reply by EliRibble** [reply]
I tried setting `TRUSTED_PROXY="*"` via environment variable on case that was needed by the TrustedProxy middleware. Also no joy.

**Comment by EliRibble** [comment, marked as answer]
I read a bunch more code and got it sorted out. In `app/Http/Middleware/TrustedProxies.php`:

```
if ('**' == $trustedProxies) {
  $this->proxies = '**';
}
```

Looks like I needed to set `TRUSTED_PROXY` to `**` not `*`. Works now.

**Reply by JC5** [reply]
I've added some details to the documentation.

https://docs.firefly-iii.org/csv/install/self_hosted/#reverse-proxies

I must stress, the correct environment variable is `TRUSTED_PROXIES`.
